### PR TITLE
Download option IIM

### PR DESCRIPTION
**Pull request by romakarol**
Download installer from artifactory for molecule test, have an option to download installer if a url is set when running the role.

**Story:**
https://zenhub.ibm.com/app/workspaces/automatonsws-5e9eeea1bdf6c83037d0c675/issues/spm-devops/automatons/494
**Travis test:**
https://app.travis-ci.com/github/IBM/spm-middleware/builds

### CODE REVIEW

File: roles/iim/tasks/linux_common.yml [hunk 1]
```diff
@@ -12,6 +12,16 @@
     src: "{{ iim_installer_path }}"
     dest: "{{tmp_dir.path}}/iim_installer.tar.gz"
   register: iim_zip
+  when: download_url is not defined
+
+- name: Download IIM installer
+  get_url:
+    dest: "{{ tmp_dir.path }}/agent.installer.linux.gtk.x86_64_{{ iim_agent_version }}.zip"
```
Comment (salob) [inline roles/iim/tasks/linux_common.yml:19]: shouldn't the destination for the copy and the download be the same? why is one a zip and the other a tar.gz ?

File: molecule/iim-191-centos-8/converge.yml [hunk 2]
```diff
@@ -10,3 +10,7 @@
 
   vars:
     iim_agent_version: 1.9.1001.20191112_1525
+    download_url: "{{ lookup('env','ARTIFACTORY_URL') }}/{{ lookup('env','ARTIFACTORY_REPO') }}/SoftwareInstallers"
+    download_header: {'X-JFrog-Art-Api': "{{ lookup('env','ARTIFACTORY_TOKEN' )}}"}
+    ansible_os_family: redhat
```
[the remark this answers is not in the document]
Comment (romakarol) [reply molecule/iim-191-centos-8/converge.yml:15]: To tell it which vars file to use. It looks like the tasks/main.yml doesnt actually 'gather' the OS type it needs to be passed it.

```
- name: Gather OS family variables
  include_vars: "{{ ansible_os_family | lower }}.yml"
  
  ...
  
  - name: Install IIM
  include: "{{ ansible_os_family | lower }}.yml"
  when: (linux_iim_dir.stat is defined and not linux_iim_dir.stat.exists) or (win_iim_dir.stat is defined and not win_iim_dir.stat.exists)
 ```

File: roles/iim/README.md [hunk 3]
```diff
@@ -12,6 +12,9 @@ None
 | ------------------- | ----------------------------------------------------- |
 | `iim_agent_version` | `1.9.1001.20191112_1525`                              |
 | `iim_install_path`  | `/opt/IBM/InstallationManager`                        |
+| `download_url`            | # set this if license and installer is being downloaded from a http server|
```
[the remark this answers is not in the document]
Comment (romakarol) [reply roles/iim/README.md:15]: There is a shell command in the tasks to activate a license which points at the installer, so I assume it to be packaged:
```
- name: Run IIM installer
  command:
    cmd: ./installc -acceptLicense -iD {{ iim_install_path }}
```

File: roles/iim/tasks/linux_common.yml (hunk 1 above)
Comment (romakarol) [reply roles/iim/tasks/linux_common.yml:19]: As you figured out the installer script saves it as a tar.gz for some reason. Im removing IIM from the script, and for the copy task I'll change it to assume .zip.

File: roles/iim/vars/redhat.yml [hunk 4]
```diff
@@ -1,2 +1,3 @@
 ---
+# url relative OR local
```
Comment (salob) [inline roles/iim/vars/redhat.yml:2]: no this is where iim gets installed, might be confusing this var with "iim_installer_path"

Comment (romakarol) [reply roles/iim/vars/redhat.yml:2]: correct, confused the two commend moved above `iim_installer_path: /tmp/iim/` in defaults/main.yml

File: molecule/iim-191-centos-8/converge.yml [hunk 5]
```diff
@@ -10,3 +10,7 @@
...
+    iim_installer_path: /IIM/agent.installer.linux.gtk.x86_64_{{ iim_agent_version }}.zip
```
[the remark this answers is not in the document]
Comment (romakarol) [reply molecule/iim-191-centos-8/converge.yml:16]: default/main.yml the default is `iim_installer_path: /tmp/iim/` .
If you want to use the download option you need to pass in a url here.

It is then used for either of:
```
- name: Copy installer
  copy:
    src: "{{ iim_installer_path }}"
```

```
- name: Download IIM installer
  get_url:
    dest: "{{ tmp_dir.path }}/agent.installer.linux.gtk.x86_64_{{ iim_agent_version }}.zip"
    url: "{{ download_url }}/{{ iim_installer_path }}"
    mode: '644'
    headers: "{{ download_header }}"
  register: iim_zip
  when: download_url is defined
```

I set it up this way based on your diagram on zenhub, is this not the desired approach? 
image

I should add this var to the README however if using this way

File: molecule/iim-191-centos-8/converge.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (salob) [reply molecule/iim-191-centos-8/converge.yml:15]: that should be determined during the build as it was before

**Pull request closed without merging**